go/consensus/tendermint/roothash/roothash.go

```diff
@@ -413,182 +422,163 @@ func (tb *tendermintBackend) onEventDataNewBlock(
+	if err != nil {
+		tb.logger.Error("error processing tendermint roothash events", "err", err)
 	}
+	tb.processEvents(ctx, events, blockHistory)
+}
 
-	_, _ = tb.onABCIEvents(ctx, events, tx.Height, true, blockHistory)
+func (tb *tendermintBackend) processEvents(ctx context.Context, events []api.Event, blockHistory map[common.Namespace]api.BlockHistory) {
+	for _, evp := range events {
+		// Redeclare variable on every iteration since the pointer to it gets broadcasted.
+		ev := evp
+
+		// Notify non-finalzed events.
+		if ev.FinalizedEvent == nil {
+			notifiers := tb.getRuntimeNotifiers(ev.RuntimeID)
+			notifiers.eventNotifier.Broadcast(&ev)
+			continue
+		}
+
+		// Process finalized event.
+		blk, value, err := tb.getBlockFromFinalizedTag(tb.ctx, ev.FinalizedEvent.Value, ev.Height)
+		if err != nil {
+			tb.logger.Error("worker: failed to get block from tag",
+				"err", err,
+			)
+			continue
+		}
+
+		notifiers := tb.getRuntimeNotifiers(value.ID)
+		// Ensure latest block is set.
+		notifiers.Lock()
+		notifiers.lastBlock = blk
+		notifiers.lastBlockHeight = ev.Height
+		notifiers.Unlock()
+
+		annBlk := &api.AnnotatedBlock{
+			Height: ev.Height,
+			Block:  blk,
+		}
+		// Commit the block to history if needed.
+		if bh, ok := blockHistory[value.ID]; ok {
+			crash.Here(crashPointBlockBeforeIndex)
+
+			err = bh.Commit(annBlk)
+			if err != nil {
+				tb.logger.Error("failed to commit block to history keeper",
+					"err", err,
+					"runtime_id", value.ID,
+					"height", ev.Height,
+					"round", blk.Header.Round,
+				)
+				// Panic as otherwise the history would become out of sync with
+				// what was emitted from the roothash backend. The only reason
+				// why something like this could happen is a problem with the
+				// history database.
+				panic("roothash: failed to index block")
+			}
+		}
+		tb.allBlockNotifier.Broadcast(blk)
+		notifiers.blockNotifier.Broadcast(annBlk)
+	}
 }
 
-func (tb *tendermintBackend) onABCIEvents(
-	ctx context.Context,
-	tmEvents []tmapi.EventWithHash,
+// EventsFromTendermint extracts staking events from tendermint events.
+func EventsFromTendermint(
+	tx tmtypes.Tx,
 	height int64,
-	doBroadcast bool,
-	blockHistory map[common.Namespace]api.BlockHistory,
+	tmEvents []tmabcitypes.Event,
+	includeFinalized bool,
 ) ([]api.Event, error) {
+	var txHash hash.Hash
+	switch tx {
+	case nil:
+		txHash.Empty()
+	default:
+		txHash = hash.NewFromBytes(tx)
+	}
+
 	var events []api.Event
+	var errs error
 	for _, tmEv := range tmEvents {
 		// Ignore events that don't relate to the roothash app.
 		if tmEv.GetType() != app.EventType {
 			continue
 		}
 
-		eh := tmEv.TxHash
-
 		for _, pair := range tmEv.GetAttributes() {
 			key := pair.GetKey()
 			val := pair.GetValue()
 
 			switch {
 			case bytes.Equal(key, app.KeyFinalized):
-				// A runtime block has been finalized.
-				if !doBroadcast {
-					// Ignore finalization event when querying events in GetEvents.
+				if !includeFinalized {
```

Comment:
I don't think this flag is needed now that processing is separate? I think the only reason for it was to avoid triggering the special processing of finalized events. Otherwise I think it is fine to include finalization events in `GetEvents`.

Reply:
yeah indeed it was now only here to not include finalization events in `GetEvents`, will remove it